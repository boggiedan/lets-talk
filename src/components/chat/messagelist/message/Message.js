import React, { useContext } from "react";
import PropTypes from "prop-types";
import { Typography } from "@material-ui/core";
import classes from "./Message.module.scss";
import classNames from "class-names";
import { UserContext } from "../../../common/usercontext/UserContext";
import format from "date-fns/format";
import Linkify from "react-linkify";
import EmbededLinkList from "./embeddedlinklist/EmbeddedLinkList";
import ImgLink from "./embeddedimg/EmbeddedImg";
import EmbeddedYoutube from "./embeddedyoutube/EmbeddedYoutube";
import { getThemedClass } from "../../../../theme/themeUtils";

const matchImg = new RegExp("(http)?s?:?(//[^\"']*.(?:png|jpg|jpeg|gif))");
const matchYoutube =
  "(?:youtube\\.com\\/\\S*(?:(?:\\/e(?:mbed))?\\/|watch\\?(?:\\S*?&?v\\=))|youtu\\.be\\/)([a-zA-Z0-9_-]{6,11})";

const Message = ({ username, userId, message, datetime }) => {
  const user = useContext(UserContext);
  const theme = user.settings.interfaceColor;

  const isCurrentUser = user.id === userId;
  const clockFormat = user.settings.clockDisplay === 12 ? "hh:mm a" : "HH:MM";
  const imgLinks = message.match(matchImg);
  const youtubeLinks = message.match(matchYoutube);

  const messageContainer = classNames(classes.messageContainer, {
    [classes.messageContainerLeft]: !isCurrentUser,
    [classes.messageContainerRight]: isCurrentUser
  });

  const messageClasses = classNames(classes.message, {
    [getThemedClass(classes, "messageLeft", theme)]: !isCurrentUser,
    [getThemedClass(classes, "messageRight", theme)]: isCurrentUser
  });

  return (
    <div className={classes.container}>
      <div className={messageContainer}>
        <div className={classes.messageWrapper}>
          {!isCurrentUser && (
            <Typography
              className={getThemedClass(classes, "messageInfo", theme)}
            >
              {username} - {format(new Date(datetime), clockFormat)}
            </Typography>
          )}
          <div className={messageClasses}>
            <Linkify>
              <Typography>{message}</Typography>
            </Linkify>
            {imgLinks && (
              <EmbededLinkList
                links={imgLinks[0].split(" ")}
                component={ImgLink}
              />
            )}
            {youtubeLinks && (
              <EmbededLinkList
                links={youtubeLinks[1].split(" ")}
                component={EmbeddedYoutube}
              />
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

Message.propTypes = {
  username: PropTypes.string.isRequired,

  userId: PropTypes.string.isRequired,

  message: PropTypes.string.isRequired,

  datetime: PropTypes.string.isRequired
};

export default Message;
