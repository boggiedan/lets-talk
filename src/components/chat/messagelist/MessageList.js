import React, { useEffect, useRef } from "react";
import PropTypes from "prop-types";
import Message from "./message/Message";
import classes from "./MessageList.module.scss";

const MessageList = ({ messages }) => {
  const messagesBottomEl = useRef(null);

  useEffect(() => scrollToBottom(), [messages]);

  const scrollToBottom = () => {
    messagesBottomEl.current.scrollIntoView({
      block: "end",
      inline: "nearest",
      behavior: "smooth"
    });
  };

  const renderMessage = item => {
    const { id, value, username, userId, datetime } = item;

    return (
      <Message
        key={id}
        username={username}
        userId={userId}
        message={value}
        datetime={datetime}
      />
    );
  };

  return (
    <section className={classes.container}>
      {messages.map(renderMessage)}
      <span ref={messagesBottomEl} />
    </section>
  );
};

MessageList.propTypes = {
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,

      value: PropTypes.string.isRequired,

      username: PropTypes.string.isRequired,

      datetime: PropTypes.string.isRequired,

      userId: PropTypes.string.isRequired
    })
  ).isRequired
};

export default MessageList;
