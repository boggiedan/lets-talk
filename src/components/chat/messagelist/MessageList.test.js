import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import MessageList from "./MessageList";
import Message from "./message/Message";

describe("Test MessageList component", () => {
  let component;

  const messages = [
    {
      id: "id",
      value: "value",
      username: "username",
      datetime: "datetime",
      userId: "userId"
    }
  ];

  it("should render match snapshot", () => {
    component = shallow(<MessageList messages={[]} />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render messages", () => {
    component = shallow(<MessageList messages={messages} />);
    expect(component.find(Message).length).toEqual(1);
  });
});
