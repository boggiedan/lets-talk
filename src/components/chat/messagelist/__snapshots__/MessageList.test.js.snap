// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MessageList component should render match snapshot 1`] = `
<section
  className="container"
>
  <span />
</section>
`;
